### ISSUE TITLE
Duplicate Key error while migrating from RethinkDB to PostgreSQL

### ISSUE THREAD

**Issue by svenseeberg** [ParabolInc/parabol#9243]
## Issue - Bug

- **How serious is the bug?** P2
- **OS:** Linux
- **Browser:** None
- **Node version:** v18.18.2
- **NPM version:** 9.8.1
- **RethinkDB version:** rethinkdb 2.4.3~0bullseye (x86_64-linux-gnu) (GCC 10.2.1)

When running `yarn pg:migrate up`, the migration fails with
```
[...]
### MIGRATION 1622680325173_threads (UP) ###
INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1622680325173_threads', NOW());

failed to create new meeting index createdAt
migrating meeting # 0
error inserting discussions error: duplicate key value violates unique constraint "Discussion_pkey"
    at Parser.parseErrorMessage (/opt/parabol/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/opt/parabol/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/opt/parabol/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/opt/parabol/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (node:events:517:28)
    at addChunk (node:internal/streams/readable:335:12)
    at readableAddChunk (node:internal/streams/readable:308:9)
    at Socket.Readable.push (node:internal/streams/readable:245:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
  length: 204,
  severity: 'ERROR',
  code: '23505',
  detail: 'Key (id)=(sroVwW3YST) already exists.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'Discussion',
  column: undefined,
  dataType: undefined,
  constraint: 'Discussion_pkey',
  file: 'nbtinsert.c',
  line: '649',
  routine: '_bt_check_unique'
}
```

I checked for duplicate keys in the RethinkDB but could not find any. I can remove the content from some RethinkDB tables which will allow the migration to continue. But that in turn results in other inconsistencies in later migration steps. I also grepped for the duplicate key in a `rethinkdb dump` file and could not find it as well. I'm currently unable to tell whether the cause is an inconsistent database or the migrations.

### Acceptance Criteria (optional)
Users can:
   - Migrate the database to install current versions of Parabol

Triage is performed to find the root cause of the bug, timeboxed to ~1 hour. 

**Estimated effort:** 1 hour to triage. More if root cause is already identified.

**Comment by Dschoordsch** [comment]
@svenseeberg the failing migration is from June 2021, could you give more detail on how you're running the migrations? To me it seems like an error to run such an old migration on existing data, unless you have a 2 year old installation which was never updated.

**Comment by Dschoordsch** [comment]
I think I know what the issue is: We used to run first all of rethink db migrations followed by all postgres migrations. This was fine going from release to release, but it fails when running a lot of them as then they might be partially out of order. This happens because some migrations move data from rethinkdb to postgres. A couple month ago (#7720) we stopped using rethink db migrations for this reason, but that doesn't help your case.

What I propose as a one off solution:
* on your target version run `yarn` to install all dependencies
* determine the order of rethinkdb and postgres migrations based on filename (rethinkdb is yyyymmddHHMM, postgres is some variation of a unix timestamp)
* manually run the migrations in the right order. You can check for interdependencies by grepping for rethink in the postgres migrations folder to see interdependencies to run larger groups of migrations.
* to run only a few rethinkdb migrations, you can either move/delete the ones you don't want to run now, or you have to modify https://github.com/ParabolInc/parabol/blob/d9b6554dce372e82a4ef9ba1b4db77d9f4ae1285/scripts/migrate.js#L28 to only run a subset with `yarn db:migrate`
* for postgrest you can do `yarn pg:migrate up <n>` with `<n>` being the number of migrations you want to run now.

Alternatively you could also jump versions in smaller increments. As the failing migration was from June 2021, I would start with [6.19.0 from July 2021](https://github.com/ParabolInc/parabol/blob/master/CHANGELOG.md#6190-2021-jul-01) and then jump only a couple of month at a time.

**Comment by svenseeberg** [comment]
> Alternatively you could also jump versions in smaller increments. As the failing migration was from June 2021, I would start with [6.19.0 from July 2021](https://github.com/ParabolInc/parabol/blob/master/CHANGELOG.md#6190-2021-jul-01) and then jump only a couple of month at a time.

I tried that but started with a later version (based on the first migrations that came in for RethinkDB), which did not work in the end. The PostgresSQL database was only populated with default and not migrated data. Maybe I'll try it starting from 6.19 again.

**Comment by Dschoordsch** [comment]
> Yes, indeed. The database has not been upgraded for quite some time. And I definitely do not want to defend that ;-)
> 
> I'm running the migrations while moving the RethinkDB database from an old to a new server. The new sever hosts the current Parabol release. After the database is imported, I'm running the following commands:
> 
> ```
> yarn
> yarn pg:build
> yarn build
> yarn db:migrate
> yarn run v1.22.21
> $ node scripts/migrate.js
> [migrate-rethinkdb]  ↑  up  ↑  20210301164434 renameFocusedPhaseItemIdToFocusedPromptId
> [migrate-rethinkdb]  ↑  up  ↑  20210324160018 mergeDuplicateUsers
> [migrate-rethinkdb]  ↑  up  ↑  20210324160029 prepareBackfillUsers
> [migrate-rethinkdb]  ↑  up  ↑  20210324160039 prepareBackfillTeams
> [migrate-rethinkdb]  ↑  up  ↑  20210324160118 swapTatorIdWithCreatedBy
> [migrate-rethinkdb]  ↑  up  ↑  20210324160125 pokerSpectate
> [migrate-rethinkdb]  ↑  up  ↑  20210330163028 addPriorityScale
> [migrate-rethinkdb]  ↑  up  ↑  20210425123805 enforceCharLimitOnUserAndTeam
> [migrate-rethinkdb]  ↑  up  ↑  20210602120505 addAdditionalRetroTemplates
> [migrate-rethinkdb]  ↑  up  ↑  20210624120419 addPokerStats
> [migrate-rethinkdb]  ↑  up  ↑  20210624121905 sortJiraDimensionFields
> [migrate-rethinkdb]  ↑  up  ↑  20210629164700 task-hash
> [migrate-rethinkdb]  ↑  up  ↑  20210708185500 removePromptTitle
> [migrate-rethinkdb]  ↑  up  ↑  20210714160201 addIsWatched
> [migrate-rethinkdb]  ↑  up  ↑  20210719161718 deletedUserEmailBackfill
> [migrate-rethinkdb]  ↑  up  ↑  20210721151034 integration-hash
> [migrate-rethinkdb]  ↑  up  ↑  20210726140543 estimate-stage-tasks
> migrating meeting # 0
> [migrate-rethinkdb]  ↑  up  ↑  20210817223727 removePromptTemplateId
> [migrate-rethinkdb]  ↑  up  ↑  20211025134235 estimateStageTaskIdBackfill
> migrating meeting # 0
> [migrate-rethinkdb]  ↑  up  ↑  20220112175838 addMissingPlainTextContentToTask
> Missed tasks found: 0
> [migrate-rethinkdb]  ↑  up  ↑  20220124600507 addMoreRetroTemplates
> [migrate-rethinkdb]  ↑  up  ↑  20220222141104 updateThreeLittlePigsPromptColors
> [migrate-rethinkdb]  ↑  up  ↑  20220329101759 dropTeams
> [migrate-rethinkdb]  ↑  up  ↑  20220524131406 addPromptToTeamPrompt
> [migrate-rethinkdb]  ↑  up  ↑  20220602164119 fixRetroTemplates
> [migrate-rethinkdb]  ↑  up  ↑  20220610093935 fixRetroPromptIds
> [migrate-rethinkdb]  ↑  up  ↑  20220614155505 addTeamCharterTemplate
> [migrate-rethinkdb]  ↑  up  ↑  20220815156512 addScrumValuesTemplate
> [migrate-rethinkdb]  ↑  up  ↑  20220815231343 addDisableAnonymity
> [migrate-rethinkdb]  ↑  up  ↑  20220819072356 addMeetingSeriesIdIndex
> [migrate-rethinkdb]  ↑  up  ↑  20220819183021 createScheduledMeetingEndIndices
> [migrate-rethinkdb]  ↑  up  ↑  20221012142324 fixDakiMigration
> [migrate-rethinkdb]  ↑  up  ↑  20221021163318 rename-christmas-template
> [migrate-rethinkdb]  ↑  up  ↑  20221025161717 addIsFreeMeetingTemplate
> [migrate-rethinkdb]  ↑  up  ↑  20221107145022 makePokerPublicTemplatesFree
> [migrate-rethinkdb]  ↑  up  ↑  20221107180234 adjustUserTiers
> [migrate-rethinkdb]  ↑  up  ↑  20221201121536 addHeardSeenRespectedTemplate
> [migrate-rethinkdb]  ↑  up  ↑  20221220161224 updateTier
> [migrate-rethinkdb]  ↑  up  ↑  20230116165620 fixStarfishTemplate
> [migrate-rethinkdb] Migration Successful
> 
> yarn pg:migrate up
> yarn run v1.22.21
> $ node-pg-migrate -f ./packages/server/postgres/pgmConfig.js up
> > Migrating files:
> > - 1611342655964_addOrganizationUserAuditTable
> > - 1613005669863_template-refs
> > - 1614030642692_refs-for-meetings
> > - 1614361628531_github-auth
> > - 1616623767058_addUserTable
> > - 1616623797090_addArrAppendFn
> > - 1616623804492_addArrDiffFn
> > - 1616623815919_addTeamTable
> > - 1616623826693_addArrMergeFn
> > - 1616623826694_add-github-search-queries
> > - 1617206728233_githubSearchQueries
> > - 1619383311996_dropUserEmailUniqueConstraint
> > - 1620834940569_lock-team
> > - 1621283671444_gh-scope
> > - 1622680325173_threads
> > - 1623102903842_discuss
> > - 1625576993064_sortJiraDimensionFields
> > - 1625576993065_addAtlassianAuthTable
> > - 1626296172790_addIsWatched
> > - 1626823580960_taskEstimate
> > - 1627394777775_addPollsTable
> > - 1631040185256_uniqueEmail
> > - 1631553388387_add-template-teams-users
> > - 1631822263196_GitHubDimensionFieldMap
> > - 1632333793178_addStripeQuantityMismatchLoggingTable
> > - 1632356828133_taskEstimateGitHub
> > - 1634351004998_addMattermostAuth
> > - 1638170546562_userDomainField
> > - 1638368905980_userUpdatedAtTrigger
> > - 1639354572171_addIntegrationProviderAndToken
> > - 1640229261909_mattermostToIntegrationsTables
> > - 1641581237527_addNonNullLastSeenAtConstraint
> > - 1642094555266_addJiraServerIntegrationProvider
> > - 1643101068628_addOAuth1TeamMemberIntegrationAuth
> > - 1644936872423_addTeamPromptResponseTable
> > - 1647259140753_dropMattermostAuthTable
> > - 1648135135375_addAzureDevOpsIntegrationProvider
> > - 1649750150687_GitLabDimensionFieldMap
> > - 1650027588544_addMeetingUserIdUniqueConstraintsToTeamPromptResponseTable
> > - 1650768593875_add-AzureDevOpsDimensionFieldMap
> > - 1650935732791_taskEstimateAzureDevOps
> > - 1651778831990_addMSTeamsIntegrationProvider
> > - 1652081624156_addJiraServerDimensionFieldMap
> > - 1652361575188_addUpdatedAtTrigger
> > - 1653315561723_addExpiresAtToTeamMemberIntegrationAuth
> > - 1654861495447_addIntegrationSearchQuery
> > - 1654861495448_OrganizationApprovedDomains
> > - 1654861495449_domainIndex
> > - 1654861495450_migrateSecureDomain
> > - 1656389795923_addAzureDevOpsDimensionFieldMapWorkItemType
> > - 1658834062488_createMeetingSeries
> > - 1659973034251_alterMeetingSeries
> > - 1663254566765_addSendSummaryEmailToUser
> > - 1663850062796_JiraDimensionFieldMap
> > - 1666387804873_userPatient0
> > - 1670328109485_addSummaryToDiscussion
> > - 1670438374119_addAIUser
> > - 1671536171066_updateTier
> > - 1673613407246_updateAIUserPic
> > - 1673629660481_addQualAIMeetingsCount
> > - 1674074684762_deleteEmptyGroups
> > - 1674237759799_meetingTemplates
> > - 1676020597201_scheduledJobOrgIdIndex
> > - 1676299334703_addSeasonalTemplates
> > - 1676341596550_updateOrgIdOnSAML
> > - 1676461074304_NotificationOrgIdIndex
> > - 1677272969994_meetingTemplatesMove
> > - 1677583911773_stripeQuantityMismatchLoggingNullableUserId
> > - 1680865284180_addDomainJoinRequestTable
> > - 1682541572157_deleteUserRethink
> > - 1682541572158_templates4
> > - 1682541573158_templates-primed
> > - 1683902222006_domainJoinRequestAutoIncrementId
> > - 1684250022646_teamHealthPhase
> > - 1685721573097_addPrePostMortemTemplates
> > - 1686226640890_addTeamHealthToAllMeetingSettingsRetrospective
> > - 1686226641890_fixed-activities
> > - 1687822236394_activity-illustrations
> > - 1688980408626_useIndexForTeamHealthAnswers
> > - 1689031398651_del-team-meetingTemplate
> > - 1689183718138_RetroReflectionGroup-part1-write
> > - 1689705702984_addDiscussionPromptQuestionsToRetroReflectionGroup
> > - 1689949187848_addTeamInsights
> > - 1690364653869_addOneOnOneMeetingType
> > - 1690544776592_addGcalIntegrationProvider
> > - 1691421540088_updateRetroReflectionGroupSortOrderType
> > - 1691422128820_fixOrganizationTier
> > - 1692638249976_isOneOnOneTeam
> > - 1692650094687_TeamMeetingTemplate
> > - 1693991480688_truncateReflectPromptIds
> > - 1694191002164_migrateSAML
> > - 1695293482618_addFreemailDomain
> > - 1695395426183_addAutoJoinToTeam
> > - 1696275124427_samlMetadataURL
> > - 1696443115482_addCustomerFeedbackTemplate
> > - 1696616922662_addPeerReviewTemplates
> > - 1697477646986_swot-typo
> > - 1698264242460_renameUserSegmentIdToPseudoId
> > - 1698265572466_renameEmailVerificationSegmentIdToPseudoId
> > - 1698265600000_addPrioritizationTemplates
> > - 1698831891828_moreFreeRetroTemplates
> ### MIGRATION 1611342655964_addOrganizationUserAuditTable (UP) ###
> 
>     DO $$
>     BEGIN
>       IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'OrganizationUserAuditEventTypeEnum') THEN
>         CREATE TYPE "OrganizationUserAuditEventTypeEnum" AS ENUM (
>           'added',
>           'activated',
>           'inactivated',
>           'removed'
>         );
>       END IF;
>       CREATE TABLE IF NOT EXISTS "OrganizationUserAudit" (
>         id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
>         "orgId" VARCHAR(100) NOT NULL,
>         "userId" VARCHAR(100) NOT NULL,
>         "eventDate" TIMESTAMP NOT NULL,
>         "eventType" "OrganizationUserAuditEventTypeEnum" NOT NULL
>       );
>       CREATE INDEX IF NOT EXISTS "idx_OrganizationUserAudit_orgId" ON "OrganizationUserAudit"("orgId");
>     END
>     $$;
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1611342655964_addOrganizationUserAuditTable', NOW());
> 
> 
> ### MIGRATION 1613005669863_template-refs (UP) ###
> #> WARNING: Need to break single transaction! <
> COMMIT;
> 
>     CREATE TABLE IF NOT EXISTS "TemplateScaleRef" (
>       "id" CHAR(24) PRIMARY KEY,
>       "scale" JSONB NOT NULL,
>       "createdAt" TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
>     );
>     CREATE TABLE IF NOT EXISTS "TemplateRef" (
>       "id" CHAR(24) PRIMARY KEY,
>       "template" JSONB NOT NULL,
>       "createdAt" TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
>     );
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1613005669863_template-refs', NOW());
> BEGIN;
> 
> 
> ### MIGRATION 1614030642692_refs-for-meetings (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1614030642692_refs-for-meetings', NOW());
> 
> 
> ### MIGRATION 1614361628531_github-auth (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1614361628531_github-auth', NOW());
> 
> 
> ### MIGRATION 1616623767058_addUserTable (UP) ###
> #> WARNING: Need to break single transaction! <
> COMMIT;
> 
>     DO $$
>     BEGIN
>       IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'TierEnum') THEN
>         CREATE TYPE "TierEnum" AS ENUM (
>           'personal',
>           'pro',
>           'enterprise'
>         );
>       END IF;
>       IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'AuthTokenRole') THEN
>         CREATE TYPE "AuthTokenRole" AS ENUM (
>           'su'
>         );
>       END IF;
>       CREATE EXTENSION IF NOT EXISTS citext;
>       CREATE TABLE IF NOT EXISTS "User" (
>         id VARCHAR(100) PRIMARY KEY,
>         email "citext" UNIQUE NOT NULL,
>         "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
>         "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
>         "inactive" BOOLEAN NOT NULL DEFAULT FALSE,
>         "lastSeenAt" TIMESTAMP WITH TIME ZONE,
>         "preferredName" VARCHAR(100) NOT NULL,
>         tier "TierEnum" NOT NULL DEFAULT 'personal',
>         picture text NOT NULL,
>         tms VARCHAR(100)[] NOT NULL DEFAULT '{}',
>         "featureFlags" VARCHAR(50)[] NOT NULL DEFAULT '{}',
>         identities JSONB[] NOT NULL DEFAULT '{}',
>         "lastSeenAtURLs" text[],
>         "segmentId" VARCHAR(100),
>         "newFeatureId" VARCHAR(100),
>         "overLimitCopy" VARCHAR(500),
>         "isRemoved" BOOLEAN NOT NULL DEFAULT FALSE,
>         "reasonRemoved" VARCHAR(2000),
>         rol "AuthTokenRole",
>         "payLaterClickCount" SMALLINT NOT NULL DEFAULT 0
>       );
>   END
>   $$;
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1616623767058_addUserTable', NOW());
> BEGIN;
> 
> 
> ### MIGRATION 1616623797090_addArrAppendFn (UP) ###
> 
>     CREATE OR REPLACE FUNCTION arr_append_uniq(anyarray, anyelement)
>       RETURNS anyarray LANGUAGE sql IMMUTABLE AS
>       'SELECT CASE WHEN array_position($1, $2) iS NULL THEN $1 || $2 ELSE $1 END;'
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1616623797090_addArrAppendFn', NOW());
> 
> 
> ### MIGRATION 1616623804492_addArrDiffFn (UP) ###
> 
>     CREATE OR REPLACE FUNCTION arr_diff(arr1 anyarray, arr2 anyarray)
>     RETURNS anyarray LANGUAGE sql IMMUTABLE AS $$
>       SELECT COALESCE(array_agg(el), '{}')
>       FROM UNNEST(arr1) el
>       WHERE el != all(arr2)
>     $$;
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1616623804492_addArrDiffFn', NOW());
> 
> 
> ### MIGRATION 1616623815919_addTeamTable (UP) ###
> 
>     DO $$
>     BEGIN
>       IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'MeetingTypeEnum') THEN
>         CREATE TYPE "MeetingTypeEnum" AS ENUM (
>           'action',
>           'retrospective',
>           'poker'
>         );
>       END IF;
>     END
>     $$;
> 
>     CREATE TABLE IF NOT EXISTS "Team" (
>       id VARCHAR(100) PRIMARY KEY,
>       name VARCHAR(100) NOT NULL,
>       "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
>       "createdBy" VARCHAR(100),
>       "isArchived" BOOLEAN NOT NULL DEFAULT FALSE,
>       "isPaid" BOOLEAN NOT NULL DEFAULT TRUE,
>       "jiraDimensionFields" JSONB[] NOT NULL DEFAULT '{}',
>       "lastMeetingType" "MeetingTypeEnum" NOT NULL DEFAULT 'retrospective',
>       tier "TierEnum" NOT NULL,
>       "orgId" VARCHAR(100) NOT NULL,
>       "isOnboardTeam" BOOLEAN NOT NULL DEFAULT FALSE,
>       "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
>     );
> 
>     CREATE INDEX IF NOT EXISTS "idx_Team_orgId" ON "Team"("orgId");
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1616623815919_addTeamTable', NOW());
> 
> 
> ### MIGRATION 1616623826693_addArrMergeFn (UP) ###
> 
>     CREATE OR REPLACE FUNCTION arr_merge(a1 anyarray, a2 anyarray)
>     RETURNS anyarray AS $$
>       SELECT ARRAY_AGG(a ORDER BY a) FROM (
>         SELECT DISTINCT UNNEST($1 || $2) AS a
>       ) s;
>     $$ LANGUAGE SQL STRICT;
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1616623826693_addArrMergeFn', NOW());
> 
> 
> ### MIGRATION 1616623826694_add-github-search-queries (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1616623826694_add-github-search-queries', NOW());
> 
> 
> ### MIGRATION 1617206728233_githubSearchQueries (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1617206728233_githubSearchQueries', NOW());
> 
> 
> ### MIGRATION 1619383311996_dropUserEmailUniqueConstraint (UP) ###
> 
>     ALTER TABLE "User"
>     DROP CONSTRAINT IF EXISTS "User_email_key";
>   ;
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1619383311996_dropUserEmailUniqueConstraint', NOW());
> 
> 
> ### MIGRATION 1620834940569_lock-team (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1620834940569_lock-team', NOW());
> 
> 
> ### MIGRATION 1621283671444_gh-scope (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1621283671444_gh-scope', NOW());
> 
> 
> ### MIGRATION 1622680325173_threads (UP) ###
> INSERT INTO "public"."PgMigrations" (name, run_on) VALUES ('1622680325173_threads', NOW());
> 
> 
> migrating meeting # 0
> error inserting discussions error: duplicate key value violates unique constraint "Discussion_pkey"
>     at Parser.parseErrorMessage (/opt/parabol/node_modules/pg-protocol/src/parser.ts:369:69)
>     at Parser.handlePacket (/opt/parabol/node_modules/pg-protocol/src/parser.ts:188:21)
>     at Parser.parse (/opt/parabol/node_modules/pg-protocol/src/parser.ts:103:30)
>     at Socket.<anonymous> (/opt/parabol/node_modules/pg-protocol/src/index.ts:7:48)
>     at Socket.emit (node:events:517:28)
>     at addChunk (node:internal/streams/readable:335:12)
>     at readableAddChunk (node:internal/streams/readable:308:9)
>     at Socket.Readable.push (node:internal/streams/readable:245:10)
>     at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
>   length: 204,
>   severity: 'ERROR',
>   code: '23505',
>   detail: 'Key (id)=(srxPAXlU9H) already exists.',
>   hint: undefined,
>   position: undefined,
>   internalPosition: undefined,
>   internalQuery: undefined,
>   where: undefined,
>   schema: 'public',
>   table: 'Discussion',
>   column: undefined,
>   dataType: undefined,
>   constraint: 'Discussion_pkey',
>   file: 'nbtinsert.c',
>   line: '649',
>   routine: '_bt_check_unique'
> }
> > Rolling back attempted migration ...
> error: duplicate key value violates unique constraint "Discussion_pkey"
>     at Parser.parseErrorMessage (/opt/parabol/node_modules/pg-protocol/src/parser.ts:369:69)
>     at Parser.handlePacket (/opt/parabol/node_modules/pg-protocol/src/parser.ts:188:21)
>     at Parser.parse (/opt/parabol/node_modules/pg-protocol/src/parser.ts:103:30)
>     at Socket.<anonymous> (/opt/parabol/node_modules/pg-protocol/src/index.ts:7:48)
>     at Socket.emit (node:events:517:28)
>     at addChunk (node:internal/streams/readable:335:12)
>     at readableAddChunk (node:internal/streams/readable:308:9)
>     at Socket.Readable.push (node:internal/streams/readable:245:10)
>     at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
>   length: 204,
>   severity: 'ERROR',
>   code: '23505',
>   detail: 'Key (id)=(srxPAXlU9H) already exists.',
>   hint: undefined,
>   position: undefined,
>   internalPosition: undefined,
>   internalQuery: undefined,
>   where: undefined,
>   schema: 'public',
>   table: 'Discussion',
>   column: undefined,
>   dataType: undefined,
>   constraint: 'Discussion_pkey',
>   file: 'nbtinsert.c',
>   line: '649',
>   routine: '_bt_check_unique'
> }
> error Command failed with exit code 1.
> ```
> 
> My next approach would be to install a 2022 version of Parabol and execute migrations with that.
> 
> *edit: The same errors occur with [`v6.80.2`](https://github.com/ParabolInc/parabol/releases/tag/v6.80.2) and [`v6.59.0`](https://github.com/ParabolInc/parabol/releases/tag/v6.59.0). The age of the migrations do not seem to be the problem.

You're also setting up a new postgres instance and migrate the data, don't you?

On second thought, looking at the migration, it does not seem to depend on the order but rather [`generateUID`](https://github.com/ParabolInc/parabol/blob/3ef1d024b977642b24a60680c9ae1b0bcd5d2c95/packages/server/postgres/migrations/1623102903842_discuss.ts#L76) generating duplicate keys. Do you have then environment variable `SERVER_ID` set to a valid integer in the range 0..1023?

**Comment by svenseeberg** [comment]
> You're also setting up a new postgres instance and migrate the data, don't you?

Yes, that is the intention ;-)

> On second thought, looking at the migration, it does not seem to depend on the order but rather [`generateUID`](https://github.com/ParabolInc/parabol/blob/3ef1d024b977642b24a60680c9ae1b0bcd5d2c95/packages/server/postgres/migrations/1623102903842_discuss.ts#L76) generating duplicate keys. Do you have then environment variable `SERVER_ID` set to a valid integer in the range 0..1023? (#9270)

Yes:
```
$ grep SERVER_ID .env
SERVER_ID='1'
```

**Comment by svenseeberg** [comment]
I think I got to the core of the solution: I missed the fact that `postdeploy` got renamed to `predeploy` and expected that the content itself was moved from RethinkDB to PostgreSQL in the database migration files. 

I ran the `yarn postdeploy` command after installing Node 14, checking out `v6.4.0` and running migrations. I do expect that going forward in steps of 10 minor releases and running the build, db:migrate, db:postgres up, postdeploy (later predeploy) commands will work.

Thanks for your feedback @Dschoordsch !
